pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/store/SegmentLocalFSDirectory.java

```diff
@@ -248,25 +250,26 @@ protected void load()
 
   private synchronized void loadData()
       throws IOException {
-    if (_columnIndexDirectory != null) {
-      return;
+    if (_columnIndexDirectory == null) {
```

Comment:
Suggest reverting this part. We do the check to prevent `loadData()` twice. We only need to add `if (CollectionUtils.isNotEmpty(_segmentMetadata.getStarTreeV2MetadataList()) {...}`

Reply:
The new changes continue to ensure _columnIndexDirectory (and _starTreeIndexReader) is loaded once. I separate their handling because the SegmentDirectory.Writer closes and sets _columnIndexDirectory to null, but Writer doesn't need to close and set _starTreeIndexReader to null, as it is for readonly.

Reply:
Hmm, it doesn't seem correct to me. When closing the writer, we should close the star-tree as well

Reply:
I think SegmentDirectory.close() is supposed to release everything owned by the directory, and IMO the SegmentDirectory..Writer.close() is more about to flush any changes.

Reply:
But after `writer.close()` is called, I don't think we can read the segment anymore.
The reader/writer name can be misleading as they are not independent of the segment directory. Unless I am missing something?